### PR TITLE
feat(schema): add deprecation field attribute

### PR DESCRIPTION
**Pull request by samugi**
### Summary

Add a deprecation record field attribute to identify fields that are deprecated. The attribute requires the `message` and `removal_in_version` fields to be configured, which are used to generate a warning when the deprecated field is configured (uses `kong.deprecation`). The optional `old_default` field can be used to determine whether the warning should be logged (in case the field was configured to a value other than the old default).

This PR adds the deprecation attribute to two kind of fields:
1. regular plugin fields that have been deprecated and that were not yet removed from the schema, i.e. the old queue parameters in datadog, opentelemetry, http-log, statsd
2. shorthand fields, for fields that have been removed from the plugin's schema and that only exist as shorthands, i.e.  acme, rate-limiting, response-ratelimiting

Example of a deprecated field (in schema/fields):
```
{ retry_count = {
    description = "Number of times to retry when sending data to the upstream server.",
    type = "integer",
    deprecation = {
      message = "http-log: config.retry_count no longer works, please use config.queue.max_retry_time instead",
      removal_in_version = "4.0",
      old_default = 10 }, }, },
```

Example of a deprecated field (in schema/shorthand_fields):
```
{ auth = {
  type = "string",
  len_min = 0,
  translate_backwards = {'password'},
  deprecation = {
    message = "acme: config.storage_config.redis.auth is deprecated, please use config.storage_config.redis.password instead",
    removal_in_version = "4.0", },
  func = function(value)
    return { password = value }
  end
}},
```

### Checklist

- [x] The Pull Request has tests (individual plugins have tests to verify the deprecation warnings are logged correctly)
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [x] (no) There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

KAG-3915

### CODE REVIEW

**Comment by samugi** [comment]
the second commit applies the deprecation attribute to the shorthand fields (where the actual field was removed from the schema).

**Comment by hanshuebner** [comment]
`removed_version` is confusing because it is in past tense, when it really is about the future.  Maybe `removal_in_version`?

File: changelog/unreleased/kong/plugin-schema-deprecation-record.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+message: "**Schema**: Add a deprecation field attribute to identify deprecated fields"
```
Comment (jschmid1) [inline changelog/unreleased/kong/plugin-schema-deprecation-record.yml:1]: changelog messages should be past tense

**Review by jschmid1 — COMMENTED** [review]
love it. Can we have a very simple test case to verify the presence of the field in the `/schemas/..` output?

File: changelog/unreleased/kong/plugin-schema-deprecation-record.yml (hunk 1 above)
Comment (samugi) [reply changelog/unreleased/kong/plugin-schema-deprecation-record.yml]: done

**Comment by samugi** [comment]
> love it. Can we have a very simple test case to verify the presence of the field in the `/schemas/..` output?

@jschmid1 I added a test, since the field is optional we have to test against something that we can make sure will always have a deprecated field, so I added one to the `dummy` custom plugin (let me know if you were thinking of a better way to test this though)

**Review by jschmid1 — APPROVED** [review]
lgtm

File: kong/db/schema/init.lua [hunk 2]
```diff
@@ -882,6 +884,16 @@ function Schema:validate_field(field, value)
     return nil, validation_errors.SUBSCHEMA_ABSTRACT_FIELD
   end
 
+  if field.deprecation then
+    local old_default = field.deprecation.old_default
+    local should_warn = old_default == nil
```
Comment (zhongweiy) [inline kong/db/schema/init.lua:889]: This line seems conflict with the doc in metaschema.lua "If `old_default` is set, logging is conditional to the field's value being different from the value of `old_default`.". Should it be `old_default ~= nil` or update the document like: 
"If `old_default` is not set, logging is conditional to the field's value being different from the value of `old_default`."
?

Comment (samugi) [reply kong/db/schema/init.lua:889]: I realize the comment wasn't very clear, I reworded it, could you check again? Thanks!

Comment (zhongweiy) [reply kong/db/schema/init.lua:889]: Thanks! My previous comment also does not correctly catch the meaning of the code. But your last version makes it more clear. And sorry for the confusion comment.

**Review by zhongweiy — APPROVED** [review]

**Pull request merged**